### PR TITLE
fix: prevent to load script again

### PR DESCRIPTION
**Pull request by hosseinmd**
https://github.com/callstack/repack/issues/749
### Summary
In this PR, I prevented `loadScript` to load scripts is loaded before

### Test plan
This is happening in remote federation with huge size of script and too many modules

### CODE REVIEW

**Review by jbroma — CHANGES_REQUESTED** [review]
Hey @hosseinmd, sorry for the late reply but I was unavailable most of the week.

The concept introduced in this PR looks great, we definitely want this in the next major (5.0) release!

Can you please add some tests for this scenarios? We need to document this behaviour as well as we can

I've also noticed that we're using the same variable 
`sriptsPromises` for both prefetching and loading - while in theoery these should not overlap, I would go with creating a separate variables to avoid potential collision in rare edge cases. 

We also need to address housekeeping of the promises that have already been fulfilled - we should remove the entires from `scriptsPromises` that are not relevant anymore.

Last thing that comes to my mind: perhaps we could return a promise in `prefetchScript` and `loadScript` instead of executing loadProcess again after awaiting the previous promise?

**Comment by hosseinmd** [comment]
> Can you please add some tests for this scenarios? We need to document this behaviour as well as we can

Yes, I will do it

**Comment by hosseinmd** [comment]
> I've also noticed that we're using the same variable
> `sriptsPromises` for both prefetching and loading - while in theoery these should not overlap, I would go with creating a separate variables to avoid potential collision in rare edge cases.

`loadScript` and `prefetchScript` are using same `resolveScript`. If we separate those, the same bug https://github.com/callstack/repack/issues/749 possible to be happening.

**Comment by hosseinmd** [comment]
> We also need to address housekeeping of the promises that have already been fulfilled - we should remove the entires from `scriptsPromises` that are not relevant anymore.

Ok, I will do it.

**Comment by hosseinmd** [comment]
> > I've also noticed that we're using the same variable
> > `sriptsPromises` for both prefetching and loading - while in theoery these should not overlap, I would go with creating a separate variables to avoid potential collision in rare edge cases.
> 
> `loadScript` and `prefetchScript` are using same `resolveScript`. If we separate those, the same bug #749 possible to be happening.

i added a test for that. https://github.com/callstack/repack/pull/756/commits/969d6d8819498ffe95ba797c414ffabcb4810358

**Comment by hosseinmd** [comment]
> > I've also noticed that we're using the same variable
> > `sriptsPromises` for both prefetching and loading - while in theoery these should not overlap, I would go with creating a separate variables to avoid potential collision in rare edge cases.
> 
> `loadScript` and `prefetchScript` are using same `resolveScript`. If we separate those, the same bug #749 possible to be happening.

In this https://github.com/callstack/repack/pull/756/commits/ee3a58c5f3f2d9d2a9ae44c47d3d0fe4b72fbdb6 commit loadScript will excute script if promise is for prefetch

**Comment by jbroma** [comment]
@hosseinmd can you merge newest main and resolve the conflicts? Then I will proceed with another review, thanks!

**Comment by hosseinmd** [comment]
Done

**Comment by jbroma** [comment]
> Done

Nice job! I'll try to review it later today

File: packages/repack/src/modules/ScriptManager/__tests__/Federated.test.ts
Comment (jbroma) [inline packages/repack/src/modules/ScriptManager/__tests__/Federated.test.ts:1]: can we move this to a a file with ScriptManager tests? almost all of the setup feels repeated here.

File: packages/repack/src/modules/ScriptManager/__tests__/Federated.test.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+beforeEach(() => {
+  ScriptManager.init();
+});
+
+afterEach(() => {
+  globalThis.__webpack_require__.repack.shared.scriptManager = undefined;
+});
+
+describe('Federated', () => {
+  it('should wait loadScript with same scriptId to finished ', async () => {
```
Comment (jbroma) [inline packages/repack/src/modules/ScriptManager/__tests__/Federated.test.ts:60]:
```suggestion
  it('should await loadScript with same scriptId to finish ', async () => {
```

File: packages/repack/src/modules/ScriptManager/ScriptManager.ts [hunk 2]
```diff
@@ -92,6 +92,10 @@ export class ScriptManager extends EventEmitter {
   }
 
   protected cache: Cache = {};
+  protected scriptsPromises: Record<
+    string,
+    (Promise<void> & { isPrefetch?: true }) | undefined
+  > = {};
```
Comment (jbroma) [inline packages/repack/src/modules/ScriptManager/ScriptManager.ts:98]: lets move this type to a separate line

**Review by jbroma — CHANGES_REQUESTED** [review]
lets clean this up a little bit, other than these few things, nice job!

**Comment by hosseinmd** [comment]
Done

File: packages/repack/src/modules/ScriptManager/__tests__/ScriptManager.test.ts [hunk 3]
```diff
@@ -2,7 +2,13 @@ import { Script } from '../Script';
...
 
 jest.mock('../NativeScriptManager', () => ({
-  loadScript: jest.fn(),
+  loadScript: jest.fn((locator) =>
+    locator.fetch
+      ? new Promise<void>((resolve) => {
+          setTimeout(() => resolve(), 10);
+        })
+      : Promise.resolve()
+  ),
```
Comment (jbroma) [inline packages/repack/src/modules/ScriptManager/__tests__/ScriptManager.test.ts:11]: can we only apply this mock implementation to the selected tests that we need? I think this can be achieved with `jest.spyOn`

**Review by jbroma — CHANGES_REQUESTED** [review]
all looks good, just one last thing 🙏

**Comment by hosseinmd** [comment]
Last commits rebased

**Review by jbroma — APPROVED** [review]
LGMT 🎉

**Comment by jbroma** [comment]
Thank you for your work on this @hosseinmd, this was no easy issue to tackle! 👏 🚀
